Use String#grapheme_clusters instead of String#scan

ricardotk002 opened the pull request:
### Summary

Both methods were introduced in Ruby 2.5 and are faster than scanning
unicode graphemes with `String#scan`

```
Warming up --------------------------------------
           scan(/X/)    43.127k i/100ms
   grapheme_clusters   103.348k i/100ms
Calculating -------------------------------------
           scan(/X/)    427.853k (± 2.4%) i/s -      2.156M in   5.042967s
   grapheme_clusters      1.045M (± 0.8%) i/s -      5.271M in   5.042360s

Comparison:
   grapheme_clusters:  1045353.5 i/s
           scan(/X/):   427852.8 i/s - 2.44x  (± 0.00) slower
```

Benchmark script:

```ruby
require "minitest/autorun"
require "benchmark/ips"

class BugTest < Minitest::Test
  def test_grapheme_clusters
    string = [0x0924, 0x094D, 0x0930].pack("U*") # "त्र"
    # string = [0x000D, 0x000A].pack("U*") # CRLF
    # string = "こにちわ"

    assert string.scan(/\X/) == string.grapheme_clusters

    Benchmark.ips do |x|
      x.report("scan(/\X/)") do
        string.scan(/\X/)
      end

      x.report("grapheme_clusters") do
        string.grapheme_clusters
      end

      x.compare!
    end
  end
end
```

### Other Information

`String#grapheme_clusters` had a bug with CRLF, which was fixed in Ruby
2.6: https://bugs.ruby-lang.org/issues/15337

Now that Rails requires Ruby 2.7+, it shouldn't be an issue.

I also noticed some of the documentation examples for `ActiveSupport::Multibyte::Chars#compose` and `ActiveSupport::Multibyte::Chars#grapheme_length` are wrong. I'm not sure if I should fix them too.

```Ruby
# It says
'क्षि'.mb_chars.grapheme_length # => 3

# It returns
'क्षि'.mb_chars.grapheme_length # => 2
```

ricardotk002 committed:
Use String#grapheme_clusters and String#each_grapheme_cluster

Both methods were introduced in Ruby 2.5 and are faster than scanning
unicode graphemes with String#scan

```
Warming up --------------------------------------
           scan(/X/)    43.127k i/100ms
   grapheme_clusters   103.348k i/100ms
Calculating -------------------------------------
           scan(/X/)    427.853k (± 2.4%) i/s -      2.156M in   5.042967s
   grapheme_clusters      1.045M (± 0.8%) i/s -      5.271M in   5.042360s

Comparison:
   grapheme_clusters:  1045353.5 i/s
           scan(/X/):   427852.8 i/s - 2.44x  (± 0.00) slower
```

Benchmark script:

```ruby
require "minitest/autorun"
require "benchmark/ips"

class BugTest < Minitest::Test
  def test_grapheme_clusters
    string = [0x0924, 0x094D, 0x0930].pack("U*") # "त्र"
    # string = [0x000D, 0x000A].pack("U*") # cr lf
    # string = "こにちわ"

    assert string.scan(/\X/) == string.grapheme_clusters

    Benchmark.ips do |x|
      x.report("scan(/\X/)") do
        string.scan(/\X/)
      end

      x.report("grapheme_clusters") do
        string.grapheme_clusters
      end

      x.compare!
    end
  end
end
```

String#grapheme_clusters had a bug with CRLF which was fixed in Ruby
2.6: https://bugs.ruby-lang.org/issues/15337

Now that Rails requires Ruby 2.7+, it shouldn't be an issue.

rafaelfranca commented:
> I'm not sure if I should fix them too.

Yes, please. Can you do in another commit in this PR?

ricardotk002 committed:
Fix examples for ActiveSupport::Multibyte::Chars methods [ci skip]

They show the wrong return values. These were recalculated in Ruby
2.7.1 and Rails 7.0.0.alpha.

ricardotk002 commented:
@rafaelfranca Done! I just sent a new commit that fixes the examples.

kaspth commented:
@ricardotk002 awesome, please continue to follow up with these sorts of things for where our RuboCop can’t detect them! And where RuboCop can detect more 2.7+ upgrades, PR pitches are also more than welcome 🙏
